### DISCUSSION TITLE
What is `messaging_android_notification_channel_id` used for?

### DISCUSSION THREAD

**Discussion by filipengberg** [invertase/react-native-firebase#5791] [Q&A] [answered]
I'm a little bit confused about the documentation about `messaging_android_notification_channel_id`: https://rnfirebase.io/messaging/usage#notification-channel-id

It says that any notification message sent via FCM will use a default channel. As I understand it you can override which channel FCM uses by adding `messaging_android_notification_channel_id` to firebase.json. Is that correct?

Does this mean that all notifications sent via FCM has to use the same channel? Or is it only used if I don't specify any channel in my message payload to FCM?

**Comment by mikehardy** [comment, marked as answer]
All notifications post as a result of FCM sent to firebase messaging will use the same channel, either the default or the one you configure.
If that's not what you want, you'll need to change the way your app works such that firebase messaging is used to receive FCM but is not in charge of posting notifications (read as: no notification block) and instead your messaging handlers integrate with a full-featured notification package (for example: notifee.app) and you use that to control how notifications are posted

example JSON you may send using FCM while integrated with notifee that would allow server-specified delivery channel: https://notifee.app/react-native/docs/integrations/fcm#server-creation

**Reply by filipengberg** [reply]
Thanks for the reply! Just do I dont misunderstand, say for example I set `messaging_android_notification_channel_id = general`. I also create this channel using notifee. Then I can send out notifications using the admin sdk to the channel_id ”general” and notifications will show up. so far so good right?

But if I want to use another channel for a separate notification, is it not possible to just create another channel using notifee, for example ”news” and set that channel_id in the admin sdk when sending the push? 

Is the single channel thing a limitation in this library or in the ”firebase messaging spec”? Since you can specify a channel_id in your notification payload I’m guessing its a library limitation?

I’m asking since if its possible to add maybe its worth for me to looking into trying to add that feature. Or maybe as you suggested its better to just use data messages and show local notifications using notifee?

Thanks again for taking your time!

**Reply by mikehardy** [reply]
single channel is a limitation of firebase-android-sdk
multiple channels via data-only messages trades off deliverability a bit (see https://dontkillmyapp.com for background issues that vendors have created) for more control
More than anything: you should try it, actual testing is the only way to get good at FCM and notifications in my experience

**Reply by filipengberg** [reply]
I just did some tests and in fact multiple channels seems to be supported.

I set `messaging_android_notification_channel_id = reminders`

Using notifee I set up two channels, `reminders` and `news`.

1. If I send a notification with `android_channel_id: reminders`. It shows a notification popup and it is connected to the `reminders` channel (looking at the notification settings for the notification in Android)
2. If I send `android_channel_id: news` it also shows a notification **and it will be connected to the `news` channel**. 🥳 
3. If I send `android_channel_id: blablabla` it will be treated as 1, that means it will be connected to `reminders`. So here it seems to be choosing the default channel set via `messaging_android_notification_channel_id`.

So according to my tests it does in fact support multiple channels as long as you create them using something like notifee and specify them in the payload. I also found some interesting documentation about this under "notification channels" here: https://firebase.google.com/docs/cloud-messaging/android/client#manifest

> ...If you prefer to create and use your own default channel, set default_notification_channel_id to the ID of your notification channel object as shown; **FCM will use this value whenever incoming messages do not explicitly set a notification channel**.

So maybe the documentation in react-native-firebase should be updated to reflect that it is actually possible to use other channels than specified in `messaging_android_notification_channel_id` as long as you specify them in your notification payload?

**Reply by mikehardy** [reply]
Great, glad that worked. I think the proposed update to the documentation is incorrect, strictly from a firebase perspective.
At the firebase level, if you send data-only notifications *no notification is generated*. There is no notification, so there is no notification channel.
If you integrate Notifee, then  you get all sorts of notification powers (as you observed) but they have nothing to do with firebase, it's all Notifee at that point

**Reply by filipengberg** [reply]
In my case im not sending any data message though, im sending out actual notifications. In my case Notifee is only used for creating the channels. 

But your probably right. Im just glad that it worked since from the documentation it sounded like I could only use a single channel for all notifications no matter how many channels I create. But as you said you do need something like notifee (which is great by the way 🤩) on top of react-native-firebase to get that functionality.

Thanks again for taking your time to answer!
